File: trading-core/lib/position.ml
type t = {
  symbol: Symbol.t;
  qty: float;
  avg_entry_price: Price.t;
  current_price: Price.t;
  strategy_id: Types.strategy_id;
  opened_at: Time_utils.t;
}

let unrealized_pnl pos =
  let diff = Price.sub pos.current_price pos.avg_entry_price in
  Money.of_float (diff *. pos.qty)

let market_value pos =
  Money.of_float (Price.to_float pos.current_price *. pos.qty)

let cost_basis pos =
  Money.of_float (Price.to_float pos.avg_entry_price *. pos.qty)

let pnl_pct pos =
  let entry = Price.to_float pos.avg_entry_price in
  if entry = 0.0 then 0.0
  else ((Price.to_float pos.current_price -. entry) /. entry) *. 100.0

let is_long pos = pos.qty > 0.0
let is_short pos = pos.qty < 0.0

let update_price pos new_price =
  { pos with current_price = new_price }

File: trading-core-js/trading-core/lib/position.js
// trading-core-js/trading-core/lib/position.js
import * as Trading_core__Money from "./money.js";
import * as Trading_core__Price from "./price.js";
function unrealized_pnl(pos) {
  const diff = Trading_core__Price.sub(pos.current_price, pos.avg_entry_price);
  return Trading_core__Money.of_float(diff * pos.qty);
}
function market_value(pos) {
  return Trading_core__Money.of_float(Trading_core__Price.to_float(pos.current_price) * pos.qty);
}
function cost_basis(pos) {
  return Trading_core__Money.of_float(Trading_core__Price.to_float(pos.avg_entry_price) * pos.qty);
}
function pnl_pct(pos) {
  const entry = Trading_core__Price.to_float(pos.avg_entry_price);
  if (entry === 0) {
    return 0;
  } else {
    return (Trading_core__Price.to_float(pos.current_price) - entry) / entry * 100;
  }
}
function is_long(pos) {
  return pos.qty > 0;
}
function is_short(pos) {
  return pos.qty < 0;
}
function update_price(pos, new_price) {
  return {
    symbol: pos.symbol,
    qty: pos.qty,
    avg_entry_price: pos.avg_entry_price,
    current_price: new_price,
    strategy_id: pos.strategy_id,
    opened_at: pos.opened_at
  };
}
export {
  cost_basis,
  is_long,
  is_short,
  market_value,
  pnl_pct,
  unrealized_pnl,
  update_price
};
